lib/datasource/go/index.ts

```diff
@@ -1,17 +1,71 @@
-import type { GetReleasesConfig, ReleaseResult } from '../types';
-import * as direct from './releases-direct';
-import * as goproxy from './releases-goproxy';
+import { cache } from '../../util/cache/package/decorator';
+import { Datasource } from '../datasource';
+import * as github from '../github-tags';
+import type { DigestConfig, GetReleasesConfig, ReleaseResult } from '../types';
+import { BaseGoDatasource } from './base';
+import { bitbucket } from './common';
+import { GoDirectDatasource } from './releases-direct';
+import { GoProxyDatasource } from './releases-goproxy';
 
-export { id } from './common';
+export class GoDatasource extends Datasource {
+  static readonly id = 'go';
 
-export { getDigest } from './digest';
+  constructor() {
+    super(GoDatasource.id);
+  }
 
-export const customRegistrySupport = false;
+  override readonly customRegistrySupport = false;
 
-export function getReleases(
-  config: GetReleasesConfig
-): Promise<ReleaseResult | null> {
-  return process.env.GOPROXY
-    ? goproxy.getReleases(config)
-    : direct.getReleases(config);
+  readonly goproxy = new GoProxyDatasource();
+  readonly direct = new GoDirectDatasource();
+
+  @cache({
+    namespace: `datasource-${GoDatasource.id}`,
+    key: ({ lookupName }: Partial<DigestConfig>) => `${lookupName}-digest`,
+  })
+  getReleases(config: GetReleasesConfig): Promise<ReleaseResult | null> {
+    return process.env.GOPROXY
+      ? this.goproxy.getReleases(config)
+      : this.direct.getReleases(config);
+  }
+
+  /**
+   * go.getDigest
+   *
+   * This datasource resolves a go module URL into its source repository
+   *  and then fetches the digest it if it is on GitHub.
+   *
+   * This function will:
+   *  - Determine the source URL for the module
+   *  - Call the respective getDigest in github to retrieve the commit hash
+   */
+  @cache({
+    namespace: GoDatasource.id,
+    key: ({ lookupName }: Partial<DigestConfig>) => `${lookupName}-digest`,
+  })
+  override async getDigest(
+    { lookupName }: Partial<DigestConfig>,
+    value?: string
+  ): Promise<string | null> {
+    const source = await BaseGoDatasource.getDatasource(lookupName);
+    if (!source) {
+      return null;
+    }
+
+    // ignore v0.0.0- pseudo versions that are used Go Modules - look up default branch instead
+    const tag = value && !value.startsWith('v0.0.0-2') ? value : undefined;
```

Comment:
`'v0.0.0-2'`: shouldn't this be `'v0.0.0-'`, as the comment says?

//cc @rarkins

Reply:
I guess it's looking for years which follow it, e.g. `v0.0.0-2021.....`

Reply:
I'm not super familar with Go, but Rhys added it, so I'll defer to him. See https://github.com/renovatebot/renovate/pull/3267/files#diff-67dacc71c63f241e34f992308f1a4e1296198f914c6cb9a91f642d5faeb2e5cf